qrexec/policy/parser.py

```diff
@@ -297,21 +312,38 @@ def __init__(self, token: str, *, filepath: Optional[pathlib.Path]=None,
     #       return '<{} value={!r} filepath={} lineno={}>'.format(
     #           type(self).__name__, str(self), self.filepath, self.lineno)
 
+    def __hash__(self):
+        return super().__hash__()
+
+    def __eq__(self, other):
+        """
+        Trap if we ever try to compare a UUID with a non-UUID.
+        This indicates a bug, so throw an exception.
+        """
+        if ((type(other) is not str) and
```

Comment:
Pylint (rightfully) complains about this. I don't think you even needs this check.
Anyway, you try very hard to break `__eq__` operator on uuid. Why?
There is `match` method that is used for comparing if given policy token matches given VM (source, target etc). `__eq__` operator is just that - comparing tokens on their own, not their meaning (in some cases it used to be the same, and in those cases the `match()` method used normal comparison, but this PR is supposed to update relevant matching methods). There is no expectation that `"@type:template" == "fedora-40"` or even `VMToken("@type:template") == VMToken("fedora-40")`.

Reply:
As said above: the `==` operator is _not_ supposed to answer whether given tokens describe the same object. That's what `match()` method is for. If there are places that use `==` for that, they need to be fixed to use the `match()` method. I see a few places that checks for explicit `@adminvm` or `dom0`, and at least one of them actually mean literal `dom0`, not any other representation of that token.
From what I see, (almost?) all your changes to avoid using this operator are _only_ to avoid hitting this exception. From what I see, the _only_ place were `==` usage needed to be changed was automatic conversion of `dom0` (and now `uuid:000...000`) to `@adminvm`. And yes, this single place needs a change.

Instrumenting `__eq__` might be valid development strategy, but it has no place in the final code.

Reply:
Instrumentation removed.